### PR TITLE
Implemented std::(w)string insertion and extraction for the Serialize…

### PR DESCRIPTION
**Pull request by jwillemsen**
…r. Serializer.h has to include Definitions.h to get the correct setting of DDS_HAS_WCHAR but this lead to a cyclic include, moved SequenceNumber out of Definitions.h to break this. Updated some other files to include the new SequenceNumber.h. This implements issue #672

    * dds/DCPS/SequenceNumber.h:
      Added.

    * dds/DCPS/CoherentChangeControl.h:
    * dds/DCPS/DataSampleHeader.h:
    * dds/DCPS/Definitions.h:
    * dds/DCPS/DisjointSequence.h:
    * dds/DCPS/Serializer.h:
    * dds/DCPS/Serializer.inl:
    * dds/DCPS/transport/framework/TransportHeader.h:
    * dds/DCPS/transport/framework/TransportQueueElement.h:
    * dds/DCPS/transport/rtps_udp/RtpsTransportHeader.h:
    * tests/DCPS/Serializer/SerializerTest.cpp:

### CODE REVIEW

**Comment by jwillemsen** [comment]
Refactoring changed implicit includes, probably the changes of #668 will fix this, will wait for that pr to be merged

**Comment by mitza-oci** [comment]
OK, I agree we can wait until #668 is merged.  Meanwhile please look at the diff of Seralizer.inl and update to match style guidelines.  Also, Serializer::read_string currently returns void.  If it returned the string length, you could use the version of std::string::assign that is given the length instead of having it do another scan for the terminator.

**Comment by jwillemsen** [comment]
Changed read_string to return the string length, didn't update the style of the added methods, the full file needs rework at some point for coding style

**Comment by mitza-oci** [comment]
I don't see what you mean regarding "the full file needs rework".  This is referring to Serializer.inl?  It looks OK to me.  Your diff has only 38 changed lines (including some that are blanks or just '{' or '}') so not much to change there.

**Comment by jwillemsen** [comment]
I was referring to Serializer.inl, the additions I made have the same layout as the existing part of the file, is there anything specific that you would like me to change?

File: dds/DCPS/Serializer.inl [hunk 1]
```diff
@@ -742,6 +742,20 @@ operator<<(Serializer& s, ACE_OutputCDR::from_wchar x)
 }
 
 ACE_INLINE bool
+operator<<(Serializer& s, const std::string& x)
+{
+  return s << x.c_str ();
```
Comment (mitza-oci) [inline dds/DCPS/Serializer.inl]: remove space before ()

File: dds/DCPS/Serializer.inl [hunk 2]
```diff
@@ -742,6 +742,20 @@ operator<<(Serializer& s, ACE_OutputCDR::from_wchar x)
...
+}
+
+#ifdef DDS_HAS_WCHAR
+ACE_INLINE bool
+operator<<(Serializer& s, const std::wstring& x)
+{
+  return s << x.c_str ();
```
Comment (mitza-oci) [inline dds/DCPS/Serializer.inl]: remove space before ()

File: dds/DCPS/Serializer.inl [hunk 3]
```diff
@@ -919,18 +933,38 @@ operator>>(Serializer& s, ACE_InputCDR::to_octet x)
...
-         && ((x.bound_ == 0) || (ACE_OS::strlen(x.val_) <= x.bound_));
+         && ((x.bound_ == 0) || (length <= x.bound_));
+}
+
+ACE_INLINE bool
+operator>>(Serializer& s, std::string& x)
+{
+  char *buf = 0;
+  const size_t length = s.read_string(buf);
+  x.assign (buf, length);
```
Comment (mitza-oci) [inline dds/DCPS/Serializer.inl]: remove space before ()

File: dds/DCPS/Serializer.inl [hunk 4]
```diff
@@ -919,18 +933,38 @@ operator>>(Serializer& s, ACE_InputCDR::to_octet x)
...
+  return s.good_bit();
+}
+
+#ifdef DDS_HAS_WCHAR
+ACE_INLINE bool
+operator>>(Serializer& s, std::wstring& x)
+{
+  ACE_CDR::WChar *buf = 0;
+  const size_t length = s.read_string(buf);
+  x.assign (buf, length);
```
Comment (mitza-oci) [inline dds/DCPS/Serializer.inl]: remove space before ()

File: dds/DCPS/Serializer.inl [hunk 5]
```diff
@@ -919,18 +933,38 @@ operator>>(Serializer& s, ACE_InputCDR::to_octet x)
...
+  const size_t length = s.read_string(const_cast<ACE_CDR::WChar*&>(x.val_));
   return s.good_bit()
-         && ((x.bound_ == 0) || (ACE_OS::strlen(x.val_) <= x.bound_));
+         && ((x.bound_ == 0) || (length <= x.bound_));
+}
+
+ACE_INLINE bool
+operator>>(Serializer& s, std::string& x)
+{
+  char *buf = 0;
```
Comment (mitza-oci) [inline dds/DCPS/Serializer.inl]: move * to the left of the space

File: dds/DCPS/Serializer.inl [hunk 6]
```diff
@@ -919,18 +933,38 @@ operator>>(Serializer& s, ACE_InputCDR::to_octet x)
...
+  const size_t length = s.read_string(buf);
+  x.assign (buf, length);
+  return s.good_bit();
+}
+
+#ifdef DDS_HAS_WCHAR
+ACE_INLINE bool
+operator>>(Serializer& s, std::wstring& x)
+{
+  ACE_CDR::WChar *buf = 0;
```
Comment (mitza-oci) [inline dds/DCPS/Serializer.inl]: move * to the left of the space

**Pull request merged**